src/interpreter.rs

```diff
@@ -1256,69 +1256,105 @@ impl<'source> Interpreter<'source> {
-            ));
+        let mut fcns: Vec<&Rule> = Vec::new();
+        for fcn_rule in fcns_rules {
+            fcns.push(fcn_rule);
         }
 
-        let mut args_scope = Scope::new();
-        for (idx, a) in args.iter().enumerate() {
-            let a = match a {
-                Expr::Var(s) => s.text(),
-                _ => unimplemented!("destructuring function arguments"),
+        let mut results: Vec<Value> = Vec::new();
+        for fcn_rule in fcns {
+            let (args, output_expr, bodies) = match fcn_rule {
+                Rule::Spec {
+                    head: RuleHead::Func { args, assign, .. },
+                    bodies,
+                    ..
+                } => (args, assign.as_ref().map(|a| &a.value), bodies),
+                _ => bail!("internal error not a function"),
+            };
+
+            if args.len() != params.len() {
+                return Err(span.source.error(
+                    span.line,
+                    span.col,
+                    format!(
+                        "mismatch in number of arguments. supplied {}, expected {}",
+                        params.len(),
+                        args.len()
+                    )
+                    .as_str(),
+                ));
+            }
+
+            let mut args_scope = Scope::new();
+            for (idx, a) in args.iter().enumerate() {
+                let a = match a {
+                    Expr::Var(s) => s.text(),
+                    _ => unimplemented!("destructuring function arguments"),
+                };
+                //TODO: check call in params
+                args_scope.insert(a.to_string(), self.eval_expr(&params[idx])?);
+            }
+
+            let ctx = Context {
+                key_expr: None,
+                output_expr,
+                value: Value::new_set(),
+            };
+
+            // Back up local variables of current function and empty
+            // the local variables of callee function.
+            let scopes = std::mem::take(&mut self.scopes);
+
+            // Set the arguments scope.
+            self.scopes.push(args_scope);
+            let value = self.eval_rule_bodies(ctx, span, bodies)?;
+            let result = match &value {
+                Value::Set(s) if s.len() == 1 => s.iter().next().unwrap().clone(),
+                Value::Set(s) if !s.is_empty() => {
+                    return Err(span.source.error(
+                        span.line,
+                        span.col,
+                        format!("function produced multiple outputs {value:?}").as_str(),
+                    ))
+                }
+                // If the function successfully executed, but did not return any value, then return true.
+                Value::Set(s) if s.is_empty() && output_expr.is_none() => Value::Bool(true),
+
+                // If the function execution resulted in undefined, then propagate it.
+                Value::Undefined => Value::Undefined,
```

Comment:
We could have a strategy where we don't gather Undefined values. In that case,

1. If `results` is empty, that means all the rules evaluated to `Undefined`.
2. If `results` is not empty, we need to check that all the values are the same.